plugins/extruderSwitchAtZ-ByDagoma.py

```diff
@@ -0,0 +1,100 @@
+# coding=utf-8
+#Name: Pause at (ByDagoma)
+#Info: Switch between extruder height or layer
+#Depend: GCode
+#Type: postprocess
+#Param: extruderSwitchLevelLayer(int:-1) Set a layer number
+
+__copyright__ = "Copyright (C) 2016 Dagoma.Fr - Released under terms of the AGPLv3 License"
+import re
+import os
+from Cura.util import profile
+
+def getPrintZValue(lineBlock):
+	'''
+	look for the last z value found just before (or at the same time) G1 code in the given block
+	'''
+	lastZ = -1
+	for line in lineBlock:
+		lastZ = getValue(line, 'Z', lastZ)
+		if line.startswith('G1 ') and (getValue(line, 'X', None) is not None or getValue(line, 'Y', None) is not None):
+			break
+
+	return lastZ
+
+
+def getValue(line, key, default = None):
+	if not key in line or (';' in line and line.find(key) > line.find(';')):
+		return default
+	subPart = line[line.find(key) + 1:]
+	m = re.search('^[0-9]+\.?[0-9]*', subPart)
+	if m is None:
+		return default
+	try:
+		return float(m.group(0))
+	except:
+		return default
+
+with open(filename, "r") as f:
+	lines = f.readlines()
+
+z = 0.
+x = 0.
+y = 0.
+pauseState = 0
+#state 0 system is not active until we get to a smaller layer than the last encountered layer (default at 99999) (print one at a time support.)
+#state 1 system is active and we are looking for our target layer z
+#state 2 system found the layer it need to write. We will switch extruder. state will be set to 0
+
+try:
+	extruderSwitchLevelLayer_i = int(extruderSwitchLevelLayer)
+except:
+	extruderSwitchLevelLayer_i = -1
+
+with open(filename, "w") as f:
+	lineIndex = 0
+	lastLayerIndex = 99999
+	lastG1 = ""
+	currentExtruder = ""
+	for lIndex in range(len(lines)):
+		line = lines[lIndex]
+		if line.startswith('T'):
+			currentExtruder = line.split(" ")[0]
+
+		if line.startswith(';'):
+			if line.startswith(';LAYER:'):
+				currentLayer = int(line[7:].strip()) + 1
+				# We are assuming there is always this sequence:
+				# G1 F2100 X122.906 Y58.862 E1750.80156
+				# ;LAYER:27
+				lastG1 = lines[lIndex-1].split()[-1]
+
+				if currentLayer < lastLayerIndex:
+					pauseState = 1
+
+				lastLayerIndex = currentLayer
+				if pauseState == 1:
+					if extruderSwitchLevelLayer_i > -1:
+						if lastLayerIndex == extruderSwitchLevelLayer_i:
+							pauseState = 2
+
+			f.write(line.rstrip() + os.linesep)
```

Comment:
```suggestion
			f.write(line)
```

I used this `os.linesep` because of a customer complaint working on Windows and opening his gcode file in a texte editor not supporting unix end of line... But python documentation states it is a bad pratice when writing file in text mode (https://docs.python.org/3/library/os.html#os.linesep).